Access to underlying UITextView for more configuration options.

marmelroy opened the pull request:
This allows much more control over the display of ASEditableTextNodes and closes #549

appleguy commented on AsyncDisplayKit/ASEditableTextNode.mm:
```diff
@@ -133,7 +134,7 @@ - (void)didLoad
 
   // Create and configure our text view.
-  _textKitComponents.textView = [[_ASDisabledPanUITextView alloc] initWithFrame:CGRectZero textContainer:_textKitComponents.textContainer];
+    _textKitComponents.textView = _textView;
```
Please fix indentation here

appleguy commented:
Awesome; thank you!  If accessing this property before the editable text node's view is created (basically before the node is "loaded") will trigger the creation of that view, please add a comment in the header about that side-effect.  Also worth mentioning in the comment, just to drive the point home, that the property can only be used on the main thread.  This is tricky because a lot of nodes are allocated / configured on background queues and that is a time you might want to set other properties on the UITextView, so we may need to do this more carefully to properly address the broad use cases.

Lastly, please add an accessor / getter implementation that has a main thread assertion in it so that folks aren't accidentally calling the method in the background.

marmelroy commented:
Excellent points. Pushed an update.

appleguy commented on AsyncDisplayKit/ASEditableTextNode.mm:
```diff
@@ -189,6 +191,21 @@ - (void)setOpaque:(BOOL)opaque
 #pragma mark -
+@dynamic textView;
+
+- (UITextView *)textView{
```
Brace style :(

appleguy commented on AsyncDisplayKit/ASEditableTextNode.mm:
```diff
@@ -189,6 +191,21 @@ - (void)setOpaque:(BOOL)opaque
+
+- (UITextView *)textView{
+  ASDisplayNodeAssertMainThread();
+  return _textView;
```
One last thing that worries me here: if this is called before it is initialized, it's likely because a client will be attempting to set properties on the text view in an -init method.  Say that it is on the main thread, like a UIView init directly hosting an ASEditableTextNode.  The problematic pitfall is that nil messaging will mean the calls fail silently, and the developer is left wondering why their configuration is not working...

appleguy commented:
Thanks for your improvement, @marmelroy!  Super close here, almost hit merge.  Last thing: please add some mechanism to alert developers accessing the text view before it is initialized.  I think this would be sufficient: ASDisplayNodeAssert(_textField != nil, @"ASEditableTextNode's text field has not been allocated yet — access it in -didLoad or after accessing the .view property instead")

Since we now have the main thread assertion (very important), an even better solution would probably be to force creation of the view and text field at the access time of the property if it hasn't been created.  This would make the API more convenient to use where the editable node is being directly hosted, and in the case of a regular node hierarchy, if initialization is being performed off-main then the assertion would catch it.

appleguy commented:
Sorry for the few days' delay here, busy week :).  I will be working on the project over the weekend and can merge the finalized change quickly.

marmelroy commented:
No worries :) So I fixed the brace style (oops). Added the warning for premature accessing and then replaced it with lazy creation of the textview in the accessor after the main thread assert.

markmark1 commented:
I would like to use pinch to Zoom to resize all the text in the ascollectionview is there an example or pointers on how to do thanks

On Sat, Jul 18, 2015 at 7:35 AM, marmelroy <notifications@github.com>
wrote:

> No worries :) So I fixed the brace style (oops). Added the warning for premature accessing and then replaced it with lazy creation of the textview in the accessor after the main thread assert.
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/facebook/AsyncDisplayKit/pull/550#issuecomment-122549905

appleguy commented:
@markmark1 that seems like a question that isn't at all related to this PR...feel free to ask it in the Google Group or Facebook Group for discussion, though!  There are a lot more people watching those who might have advice.  It's not totally clear the effect you're going for, because redrawing text is pretty expensive and usually has discrete font sizes (at least integers) that wouldn't feel very natural to control with a continuous gesture like a pinch.  Even a slider might be a more understandable UI.  I could be misunderstanding your idea, though.

appleguy commented:
@marmelroy thanks!  Rerunning the build to see if we can get a green one, then will merge.

appleguy commented on AsyncDisplayKit/ASEditableTextNode.mm:
```diff
@@ -77,6 +78,7 @@ - (instancetype)init
   _textKitComponents = [ASTextKitComponents componentsWithAttributedSeedString:nil textContainerSize:CGSizeZero];
   _textKitComponents.layoutManager.delegate = self;
   _wordKerner = [[ASTextNodeWordKerner alloc] init];
+  _textView = [[_ASDisabledPanUITextView alloc] initWithFrame:CGRectZero textContainer:_textKitComponents.textContainer];
```
Hmm, unfortunately this isn't right.  We can't allocate views in -init, as it is likely to be called off the main thread.  It's necessary to refactor this so that no text view initialization happens in -init.

appleguy commented on AsyncDisplayKit/ASEditableTextNode.mm:
```diff
@@ -133,7 +135,7 @@ - (void)didLoad
 
   // Create and configure our text view.
-  _textKitComponents.textView = [[_ASDisabledPanUITextView alloc] initWithFrame:CGRectZero textContainer:_textKitComponents.textContainer];
+  _textKitComponents.textView = _textView;
```
Since we're doing this, it looks like we might as well not add the _textView instance variable at all.  Let's just use _textKitComponents.textView, anywhere and everywhere that _textView would otherwise need to be used.  This ensures that we wouldn't have a case where only one of them is initialized but they aren't set up to be the same.

appleguy replied:
Let's refactor this area to do something like this:

UITextView *textView = self.textView; // The textView accessor method should set up _textKitComponents.textView

Then just use textView directly in the following six lines.  This eliminates 6 message sends as well as makes the code more readable.

appleguy commented on AsyncDisplayKit/ASEditableTextNode.h:
```diff
@@ -19,6 +19,12 @@
+  @abstract Access to underlying UITextView for more configuration options.
+  @warning This property should only be used on the main thread and should not be accessed before the editable text node's view is created.
+ */
+@property (nonatomic, readwrite, strong) UITextView *textView;
```
This should not be a readwrite property as we don't want arbitrary UITextViews to be able to assigned here, since it needs to be a custom subclass.

appleguy commented on AsyncDisplayKit/ASEditableTextNode.mm:
```diff
@@ -41,6 +41,7 @@ @interface ASEditableTextNode () <UITextViewDelegate, NSLayoutManagerDelegate>
 {
   @private
   // Configuration.
+  UITextView *_textView;
```
Let's just remove this variable completely, because a) it is not needed, and b) increases the potential for bugs where it is out of sync with the _textKitComponents.

appleguy commented on AsyncDisplayKit/ASEditableTextNode.mm:
```diff
@@ -189,6 +191,25 @@ - (void)setOpaque:(BOOL)opaque
+  return _textView;
+}
+
+- (void)setTextView:(UITextView *)textView
```
Seems like we should get rid of this since presumably the _ASDisabledPanUITextView is important.

appleguy commented on AsyncDisplayKit/ASEditableTextNode.mm:
```diff
@@ -189,6 +191,25 @@ - (void)setOpaque:(BOOL)opaque
 @synthesize delegate = _delegate;
 
 #pragma mark -
+@dynamic textView;
```
No need for either @dynamic or @synthesize, right?  I could be wrong, but with the readonly property and implementing the getter method below, I think we should be good.  Also remove the pragma mark because it doesn't seem to be adding anything (e.g. it doesn't have a name or any real value because it is so close to the other one).

appleguy commented on AsyncDisplayKit/ASEditableTextNode.mm:
```diff
@@ -189,6 +191,25 @@ - (void)setOpaque:(BOOL)opaque
+- (UITextView *)textView
+{
+  ASDisplayNodeAssertMainThread();
+  if (!_textView) {
```
This will then just be !_textKitComponents.textView,

and if it doesn't exist, just set _textKitComponents.textView = alloc / init.

Then just quickly check the various areas in the class that use _textKitComponents to ensure that in all cases, this accessor method will be called at least once before the textView is used, so that it is set up properly.

appleguy commented:
Sorry for all the comments!  This one won't require much code, but it's very careful that the threading & initialization paths are correct.

markmark1 commented:
Ok will do @appleguy

On Sat, Jul 18, 2015 at 2:17 PM, appleguy <notifications@github.com>
wrote:

> @markmark1 that seems like a question that isn't at all related to this PR...feel free to ask it in the Google Group or Facebook Group for discussion, though!  There are a lot more people watching those who might have advice.  It's not totally clear the effect you're going for, because redrawing text is pretty expensive and usually has discrete font sizes (at least integers) that wouldn't feel very natural to control with a continuous gesture like a pinch.  Even a slider might be a more understandable UI.  I could be misunderstanding your idea, though.
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/facebook/AsyncDisplayKit/pull/550#issuecomment-122597540

markmark1 commented:
Wanted to know what the right way to do it for all the text uitextviews in asdk as its super expensive  and discrete font sizes 

Hence posted in the uitextviews thread but will check the Facebook and Google groups as well

On Sat, Jul 18, 2015 at 2:17 PM, appleguy <notifications@github.com>
wrote:

> @markmark1 that seems like a question that isn't at all related to this PR...feel free to ask it in the Google Group or Facebook Group for discussion, though!  There are a lot more people watching those who might have advice.  It's not totally clear the effect you're going for, because redrawing text is pretty expensive and usually has discrete font sizes (at least integers) that wouldn't feel very natural to control with a continuous gesture like a pinch.  Even a slider might be a more understandable UI.  I could be misunderstanding your idea, though.
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/facebook/AsyncDisplayKit/pull/550#issuecomment-122597540

markmark1 commented:
@appleguy can you please let me know what the links to the facebook group and google group

On Sat, Jul 18, 2015 at 2:17 PM, appleguy <notifications@github.com>
wrote:

> @markmark1 that seems like a question that isn't at all related to this PR...feel free to ask it in the Google Group or Facebook Group for discussion, though!  There are a lot more people watching those who might have advice.  It's not totally clear the effect you're going for, because redrawing text is pretty expensive and usually has discrete font sizes (at least integers) that wouldn't feel very natural to control with a continuous gesture like a pinch.  Even a slider might be a more understandable UI.  I could be misunderstanding your idea, though.
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/facebook/AsyncDisplayKit/pull/550#issuecomment-122597540

appleguy commented:
@markmark1 —

https://groups.google.com/forum/#!forum/asyncdisplaykit
http://fb.me/paperf8

marmelroy commented:
@appleguy - I pushed a simpler and cleaner implementation this morning. Travis CI stalls during building but doesn't look like it's related to this pull request.

appleguy commented:
Something is really messed up with travis!  @eanagel I wonder if It could be any changes we made to the tests?

Restarting the build.

appleguy commented:
We are working on the build issue.  It appears to be the recent Cocoapods version change, which in fact broke the internal Pinterest build as well.  Sorry that's blocking this for now, should be fixed tomorrow.

marmelroy commented:
@appleguy looks like Travis issues were fixed!

appleguy commented:
@marmelroy this change is officially badass-ified.  Thank you for surfacing the need.  We actually also added returnKeyType here as a direct passthrough property, FWIW.
